substratevm/src/com.oracle.svm.configure/src/com/oracle/svm/configure/config/ConfigurationKind.java

```diff
@@ -0,0 +1,129 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.configure.config;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public final class ConfigurationKind {
+    private static Map<ConfigurationMemberKind, Map<ConfigurationAccessKind, ConfigurationKind>> cache = new HashMap<>();
+
+    private final ConfigurationMemberKind memberKind;
+    private final ConfigurationAccessKind accessKind;
+
+    private ConfigurationKind(ConfigurationMemberKind memberKind, ConfigurationAccessKind accessKind) {
+        this.memberKind = memberKind;
+        this.accessKind = accessKind;
+    }
+
+    public ConfigurationMemberKind getMemberKind() {
+        return memberKind;
+    }
+
+    public ConfigurationAccessKind getAccessKind() {
+        return accessKind;
+    }
+
+    public static ConfigurationKind get(ConfigurationMemberKind memberKind, ConfigurationAccessKind accessKind) {
+        return cache.computeIfAbsent(memberKind, k -> new HashMap<>()).computeIfAbsent(accessKind, k -> new ConfigurationKind(memberKind, accessKind));
+    }
+
+    public enum ConfigurationMemberKind {
+        /**
+         * The member is public and declared in the type in question.
+         */
+        DECLARED_AND_PUBLIC,
+
+        /**
+         * The member is declared in the type in question.
+         */
+        DECLARED,
+
+        /**
+         * The member is public and is either declared or inherited in the type in question.
+         */
+        PUBLIC,
+
+        /**
+         * The member is either declared or inherited in the type in question.
+         */
+        PRESENT;
+
+        private boolean isMoreSpecificThan(ConfigurationMemberKind other) {
+            return other == null || ordinal() < other.ordinal();
+        }
+
+        public ConfigurationMemberKind intersect(ConfigurationMemberKind other) {
+            if (equals(DECLARED) && PUBLIC.equals(other) || equals(PUBLIC) && DECLARED.equals(other)) {
+                return DECLARED_AND_PUBLIC;
+            }
+            return this.isMoreSpecificThan(other) ? this : other;
+        }
+
+        private ConfigurationMemberKind union(ConfigurationMemberKind other) {
+            return equals(other) ? this : PRESENT;
+        }
+
+        public boolean includes(ConfigurationMemberKind other) {
+            if (equals(DECLARED_AND_PUBLIC)) {
+                return DECLARED.equals(other) || PUBLIC.equals(other);
+            }
+            if (equals(PRESENT)) {
+                return true;
+            }
+            return equals(other);
+        }
+    }
+
+    public enum ConfigurationAccessKind {
+        NONE,
+        QUERIED,
+        ACCESSED;
```

Comment:
Would `INVOKED` fit better to what this value describes?

Reply:
I chose `ACCESSED` to have a general term that also applies to fields, since these values are attached to all member types, even if only trivially. I agree it's a bit unclear, but I think `INVOKED` is too specific here. Something like `USED` could be an alternative.